### PR TITLE
add ci

### PR DESCRIPTION
**Pull request by rtest12**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    steps:
+      - name: Git clone repo
+        uses: actions/checkout@v3
+
+      - name: Setup Node.js
+        uses: actions/setup-node@v3
+        with:
+          cache: 'npm'
+          cache-dependency-path: ./angular/package-lock.json
+
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yml:69]: Let's reduce number of copy paster.
1. Clone repo
2. Setup npm 
3. then run tests
4. if succeed - then build and push

Let's reduce the execution time to a minimum.


an idea:
```
jobs:
  docker-build:
    name: Docker build
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Build Docker image
        run: |
          docker build -t foo/bar:$GITHUB_SHA
          mkdir -p path/to/artifacts
          docker save foo/bar:$GITHUB_SHA > path/to/artifacts/docker-image.tar
          
      - name: Temporarily save Docker image
        uses: actions/upload-artifact@v2
        with:
          name: docker-artifact
          path: path/to/artifacts
          retention-days: 1

  docker-deploy:
    name: Deploy to Docker Hub
    runs-on: ubuntu-latest
    needs: docker-build
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Retrieve saved Docker image
        uses: actions/download-artifact@v2
        with:
          name: docker-artifact
          path: path/to/artifacts

      - name: Docker load
        run: |
          cd path/to/artifacts
          docker load < docker-image.tar
          # docker_build_push.sh
```
https://stackoverflow.com/questions/57498605/github-actions-share-workspace-artifacts-between-jobs

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      - name: Send mail
+        uses: dawidd6/action-send-mail@v3
+        with:
+          server_address: ${{secrets.SMTP_ADRESS}}
+          server_port: 465
+          secure: true
+          username: ${{secrets.MAIL_USERNAME}}
+          password: ${{secrets.MAIL_PASSWORD}}
+          subject: ${{ github.job }} job of ${{ github.repository }} has ${{ job.status }}
+          to: ${{secrets.MAIL_USERNAME}}
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yml:122]: should be the author of the commit, not hardcoded value from secrets

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

**Comment by dmitry-mightydevops** [comment]
@rtest12 also - do you have docker-compose.yaml to review?

**Comment by dmitry-mightydevops** [comment]
@rtest12 when you done:
- add label "Needs Review" to your PR
- re-request me as reviewer
- remove the PR from the "draft" status

Otherwise I don't know if you still have work in propgress, or finished changes requested and ready for another review

**Comment by dmitry-mightydevops** [comment]
@rtest12 you also need to re-request the review
image

click on the () icon on the left from the red icon

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -1,135 +1,132 @@
...
+              cache-dependency-path: ./angular/package-lock.json
+    
+          - name: install dependencies
+            run: npm install
+    
+          - name: run tests
+            run: npm run test -- --watch=false --browsers=ChromeHeadless
+    
+          - name: if fail
+            if: failure()
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yml:48]: this is cryptic. You need to reference the step by ID to check if it was successful or failed.

File: nginx.conf [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+server {
+    listen [::]:8080 default_server;
+
+    root /var/www/html;
+    index index.html index.htm index.nginx-debian.html;
+
+    server_name _;
+
+    location / {
+            try_files $uri $uri/ =404;
```
[the remark this answers is not in the document]
Comment (dmitry-mightydevops) [reply nginx.conf:38]: your soluton using deny all is not fixing the problem on line 38. Your security hole is still there. If in the future someone removes lines 33-35 or make error in regexp - anyone can download files there were not supposed to.

Pls read official docs how to use try_files, and what is the purpose of passing $uri $uri/ =404; args.
http://nginx.org/en/docs/http/ngx_http_core_module.html#try_files

File: .github/workflows/ci.yml (hunk 2 above)
Comment (dmitry-mightydevops) [reply .github/workflows/ci.yml]: @rtest12 I don't see this is fixed. Still `to: ${{secrets.MAIL_USERNAME}}`

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]
see my feedback

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -120,7 +121,7 @@
               username: ${{secrets.MAIL_USERNAME}}
               password: ${{secrets.MAIL_PASSWORD}}
               subject: ${{ github.job }} job of ${{ github.repository }} has ${{ job.status }}
-              to: ${{secrets.MAIL_USERNAME}}
+              to: omelchenko.sochi@gmail.com
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yml:124]: it should not be hardcoded this way either :)
see variables that github makes it available inside github action run for each build - use the author of the commit. So If I commit into your repo - I (as the author) should receive it, not you.

File: nginx.conf [hunk 6]
```diff
@@ -34,8 +34,8 @@ server {
         deny all;
     }
 
-    location / {
-            try_files $uri $uri/ =404;
+    location ~ \.(js|css|png|jpg|gif|swf|ico|pdf|mov|fla|zip|rar)$ {
+       try_files $uri =404;
     }
```
Comment (dmitry-mightydevops) [inline nginx.conf:39]: so this is a little bit closer :)  but still not perfect.

**hint**:  use index.html inside try_files and pass query args as needed.

That way you don't depend on the value in $uri and don't need regexp in location.

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

**Review by dmitry-mightydevops — APPROVED** [review]
@rtest12 
Макс - проапрувил. Сделаем на след. неделе групповой митинг с ребятами, что прошли тест и обсудим

- дату старта практики
- процесс
- что нужно будет иметь к началу практики

Марина отправит уведомление.

**Comment by rtest12** [comment]
> @rtest12 Макс - проапрувил. Сделаем на след. неделе групповой митинг с ребятами, что прошли тест и обсудим
> 
>     * дату старта практики
> 
>     * процесс
> 
>     * что нужно будет иметь к началу практики
> 
> 
> Марина отправит уведомление.

Хорошо, буду на связи.

**Pull request closed without merging**